File: cpp/dsa/BinTreeNode.cpp
#include <iostream>
#include <memory>
#include "LinkedList.cpp"

namespace tree {
template<typename T>
class BinTreeNode {
public:
    std::unique_ptr<BinTreeNode> left;
    std::unique_ptr<BinTreeNode> right;

    // Default constructor
    BinTreeNode(T inVal)
        : val{inVal}, left{nullptr}, right{nullptr}, duplicates{} {
    }

    // Move constructor
    BinTreeNode(BinTreeNode &&other) noexcept 
        : val{std::move(other.val)}, left{std::move(other.left)}, right{std::move(other.right)}, duplicates{std::move(other.duplicates)} {
        other.left = nullptr;
        other.right = nullptr;
    }

    ~BinTreeNode() = default;

    /// @brief  Return the value of the node
    /// @param  node 
    /// @return Value of node
    auto getVal(const BinTreeNode<T>* node) const -> int {
        return node->val;
    }

    /// @brief  Insert a new node into the tree with inorder traversal
    /// @param  newVal
    auto insert(const T newVal) -> const void {
        if (newVal < val) {
            if (this->left) {
                this->left->insert(newVal);
            } else {
                this->left = std::make_unique<BinTreeNode>(newVal);
            }
        } else if (newVal > val) {
            if (this->right) {
                this->right->insert(newVal);
            } else {
                this->right = std::make_unique<BinTreeNode>(newVal);
            }
        } else { // newVal == val
            duplicates.insert_end(val);
        }
    }

    auto operator+(int inc) -> BinTreeNode<T> {
        BinTreeNode<T> curNode = this;
        while(curNode) {
            if(curNode->left) {
                curNode->left;
            } else if (curNode->right) {
                curNode->right;
            }
        }
    }

    auto print_preorder() const -> void {
        std::cout << getVal(this) << " ";
        if (duplicates.head) {
            duplicates.print_list();
        }
        if (left) left->print_preorder();
        if (right) right->print_preorder();
    }

    auto print_inorder() const -> void {
        if (left) left->print_inorder();
        std::cout << getVal(this) << " ";
        if (duplicates.head) {
            duplicates.print_list();
        }
        if (right) right->print_inorder();
    }

    auto print_postorder() const -> void {
        if (left) left->print_inorder();
        if (right) right->print_inorder();
        std::cout << getVal(this) << " ";
        if (duplicates.head) {
            duplicates.print_list();
        }
    }

private:
    LinkedList<T> duplicates;
    T val;
};

template<typename T>
auto main() -> int {
    int val = 1;
    auto root = std::make_unique<BinTreeNode<int>>(val);
    root->insert(7);
    root->insert(7);
    root->insert(3);
    root->insert(9);
    root->insert(8);
    //root->left = std::make_unique<BinTreeNode<int>>(++val);
    //root->right = std::make_unique<BinTreeNode<int>>(++val);
    //root->right->left = std::make_unique<BinTreeNode<int>>(++val);
    //root->right->right = std::make_unique<BinTreeNode<int>>(BinTreeNode(++val));
    root->print_preorder();
    std::cout << std::endl;
    root->print_inorder();
    std::cout << std::endl;
    root->print_postorder();
    std::cout << std::endl;
    return 0;
}

}

int main() {
    return tree::main<int>();
}

File: python/bintreenode.py
class BinTreeNode:
    value: int
    left: "BinTreeNode | None"
    right: "BinTreeNode | None"
    def __init__(self, value=0):
        self.value = value
        self.left = None
        self.right = None

    def insert(self, new_val):
        if new_val < self.value:
            if self.left:
                self.left.insert(new_val)
            else:
                self.left = BinTreeNode(new_val)
        else:
            if self.right:
                self.right.insert(new_val)
            else:
                self.right = BinTreeNode(new_val)
    

    @staticmethod
    def print_inorder(root):
        if not root:
            return
        BinTreeNode.print_inorder(root.left)
        print(root.value)
        BinTreeNode.print_inorder(root.right)

    @staticmethod
    def print_left(root):
        if not root:
            return
        print(root.value)
        BinTreeNode.print_inorder(root.left)

    @staticmethod
    def print_right(root):
        if not root:
            return
        print(root.value)
        BinTreeNode.print_inorder(root.right)

    @staticmethod
    def levels(root) -> int:
        if not root:
            return 0
        resultL = 1 + BinTreeNode.levels(root.right)
        resultR =  1 + BinTreeNode.levels(root.left)
        return max(resultL, resultR)


def main():
    root = BinTreeNode(3)
    root.insert(0)
    root.insert(-1)
    root.insert(1)
    # root.insert(2)
    # root.insert(6)
    # root.insert(5)
    # root.insert(11)
    # root.insert(-1)
    root.insert(4)
    # root.insert(7)
    # root.insert(8)
    # root.insert(87)
    # root.insert(289)

    BinTreeNode.print_inorder(root)
    # BinTreeNode.print_left(root)
    # BinTreeNode.print_left(root)

    print(f"Tree is {BinTreeNode.levels(root)} deep.")

if __name__ == "__main__":
    main()